parasol/solr/client.py

```diff
@@ -0,0 +1,131 @@
+    core_admin_handler = 'admin/cores'
+    #: select handler
+    select_handler = 'select'
+    #: schema handler
+    schema_handler = 'schema'
+    # update handler
+    update_handler = 'update'
+    #: core or collection
+    collection = ''
+    # commitWithin definition
+    commit_within = 1000
+
+    def __init__(self, *args, **kwargs):
+        # For now, a generous init that will override valuesbut not
+        # choke on an unexpected one.
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+        # Don't allow session configuration to be so easily overwritten
+        self.session = requests.Session()
+        self.session.headers = {
+            'User-Agent': 'Parasol/%s' % parasol_version
+        }
+
+        self.schema = Schema(self)
+        self.update = Update(self, commitWithin=1000)
+        self.core_admin = CoreAdmin(self)
+
+    def build_url(self, handler):
+        '''Return a url to a handler based on core and base url'''
+        # Two step proecess to avoid quirks in urljoin behavior
+        # First, join the collection/core with a slashes appended
+        # just in case so # it doesn't ovewrite the base url
+        # (extras cause no issues)
+        collection = urljoin('%s/' % self.solr_url, '%s/' % self.collection)
+        # then return the core joined with handler -- without slash per
+        # Solr API docs.
+        return urljoin(collection, handler)
+
+    def make_request(self, meth, url, headers={},
+                      params={}, data=None, **kwargs):
+        '''Private method for making a request to Solr, wraps session.request'''
+        # always add wt=json for JSON api
+        if 'wt' not in params:
+            params.update({'wt': 'json'})
+        start = time.time()
+        response = self.session.request(
+            meth,
+            url,
+            params=params,
+            headers=headers,
+            data=json.dumps(data),
+            **kwargs)
+        # log the time as needed
+        total_time = time.time() - start
+        log_string = '%s %s=>%d: %f sec' % \
+            (meth.upper(), url, response.status_code, total_time)
+        if response.status_code == requests.codes.ok:
+            logger.debug(log_string)
+            # do further error checking on the response because Solr
+            # may return 200 but pass along its own error codes and information
+            output = AttrDict(response.json())
+            if 'responseHeader' in output \
+                    and output.responseHeader.status != 0:
+                log_string = (
+                    '%s %s (%d): %s' %
+                    meth.upper(),
+                    url,
+                    output.responseHeader.status,
+                    output.responseHeader.status
+                )
+                logger.error(log_string)
+                return None
+            return AttrDict(response.json())
+        logger.error(log_string)
```

Comment:
Error logging should provide *additional* information to the debug output. Not sure what more information we can provide here, but could at least call attention to the error code more loudly than the debug output does.

Reply:
I'll see what else I can add. I was also using error to highlight it more vigorously in the stack.